sdk/lib/rtl/rangelist.c

```diff
@@ -13,546 +13,1556 @@
+
+    Start = Maximum - (Length - 1) - ((Maximum - (Length - 1)) % Alignment);
+
+    if ((Minimum > Maximum) ||
+        ((Maximum - Minimum) < (Length - 1)) ||
+        ((Minimum + Alignment) < Minimum) ||
+        (Start < Minimum) ||
+        Length == 0 ||
+        Alignment == 0)
     {
+      #ifdef NDEBUG
+        DPRINT1("RtlFindRange: [%X] %p, %X, [%I64X-%I64X], %X, %X\n", Flags, RangeList, RangeList->Count, Minimum, Maximum, Length, Alignment);
+      #endif
+
+        DPRINT1("RtlFindRange: Testing: STATUS_INVALID_PARAMETER\n");
         return STATUS_INVALID_PARAMETER;
     }
 
-    if (IsListEmpty(&RangeList->ListHead))
-    {
-        *Start = ROUND_DOWN(Maximum - (Length - 1), Alignment);
-        return STATUS_SUCCESS;
-    }
+    End = Start + Length - 1;
+    RtlGetLastRange(RangeList, &Iterator, &TmpRange);
 
-    NextEntry = NULL;
-    Entry = RangeList->ListHead.Blink;
-    while (Entry != &RangeList->ListHead)
+    do
     {
-        CurrentEntry = CONTAINING_RECORD(Entry, RTL_RANGE_ENTRY, Entry);
-
-        RangeMax = NextEntry ? (NextEntry->Range.Start - 1) : Maximum;
-        if (RangeMax + (Length - 1) < Minimum)
+        DPRINT("RtlFindRange: Testing [%I64X-%I64X]\n", Start, End);
+
+        if (RtlpIsRangeAvailable(&Iterator,
+                                 Start,
+                                 End,
+                                 AttributeAvailableMask,
+                                 ((Flags & RTL_RANGE_LIST_SHARED_OK) != 0),
+                                 ((Flags & RTL_RANGE_LIST_NULL_CONFLICT_OK) != 0),
+                                 FALSE,
+                                 Context,
+                                 Callback))
         {
-            return STATUS_RANGE_NOT_FOUND;
-        }
-
-        RangeMin = ROUND_DOWN(RangeMax - (Length - 1), Alignment);
-        if (RangeMin < Minimum ||
-            (RangeMax - RangeMin) < (Length - 1))
-        {
-            return STATUS_RANGE_NOT_FOUND;
+            *OutStart = Start;
+            ASSERT((*OutStart >= Minimum) && (*OutStart + Length - 1 <= Maximum));
+            DPRINT("RtlFindRange: return STATUS_SUCCESS\n");
+            return STATUS_SUCCESS;
         }
 
-        DPRINT("RangeMax: %I64x\n", RangeMax);
-        DPRINT("RangeMin: %I64x\n", RangeMin);
+        Start = ((PRTLP_RANGE_LIST_ENTRY)(Iterator.Current))->Start;
+        DPRINT("RtlFindRange: Iterator Start %I64X\n", Start);
 
-        if (RangeMin > CurrentEntry->Range.End)
-        {
-            *Start = RangeMin;
-            return STATUS_SUCCESS;
+        if (Start - Length > Start) {
+            DPRINT("RtlFindRange: break\n");
+            break;
         }
 
-        NextEntry = CurrentEntry;
-        Entry = Entry->Blink;
-    }
-
-    RangeMax = NextEntry ? (NextEntry->Range.Start - 1) : Maximum;
-    if (RangeMax + (Length - 1) < Minimum)
-    {
-        return STATUS_RANGE_NOT_FOUND;
-    }
+        Start = Start - Length - (Start % Alignment);
```

Comment:
Is it only `(Start % Alignment)` ? or should it be `((Start - Length) % Alignment)` ?

Reply:
Yes, (Start - Length)